src/Simulation/CSharpGeneration/RewriteStep.fs

```diff
@@ -61,9 +61,24 @@ type Emitter() =
                 if content <> null then CompilationLoader.GeneratedFile(source, dir, ".dll.g.cs", content) |> ignore
 
             if not compilation.EntryPoints.IsEmpty then
-                let callable = context.allCallables.[Seq.exactlyOne compilation.EntryPoints]
-                let content = EntryPoint.generate context callable
-                CompilationLoader.GeneratedFile(callable.SourceFile, dir, ".EntryPoint.g.cs", content) |> ignore
+
+                let entryPointCallables =
+                    compilation.EntryPoints
+                    |> Seq.map (fun ep -> context.allCallables.[ep])
+
+                let main = EntryPoint.mainNamespace entryPointCallables
+
+                let entryPointSources =
+                    entryPointCallables
+                    |> Seq.groupBy (fun ep -> ep.SourceFile)
+
+                let (sourceFile, callables) = Seq.head entryPointSources
+                let content = EntryPoint.generateEntryPointSource context callables (Some main)
+                CompilationLoader.GeneratedFile(sourceFile, dir, ".EntryPoint.g.cs", content) |> ignore
```

Comment:
I gave this one quite a lot of thought, and I believe it creates much more work and special casing to handle the possibility that, if we have a special file for the main, we may have a collision with another file of the same name being used as an entry point file. My conclusion was to avoid having to deal with this by simply putting the main in the first (possibly only) entry point file.

Reply:
Hmm, now that you mention file collisions, I think there's already a collision if you have two Q# files `Foo.qs` and `Foo.EntryPoint.qs`, where `Foo.qs` contains the `@EntryPoint()`. `Foo.qs` generates `Foo.g.cs` and `Foo.EntryPoint.g.cs`, and `Foo.EntryPoint.qs` also generates `Foo.EntryPoint.g.cs`.

I think it would be fixed if the suffix was `.g.EntryPoint.cs` instead of `.EntryPoint.g.cs`. In that case, you could also generate a separate file for the Main method named `EntryPoint.g.Main.cs` (or something with a unique `.g.???.cs` suffix), and I don't think it could conflict with any user file.

Reply:
So for my understanding, the problem you are suggesting is that, for the two existing suffixes `.g.cs` and `.Entrypoint.g.cs`, one suffix is a substring of the other. And the solution is to have a suffix set so that no suffix is the substring of another. For example `.g.cs`, `.g.EntryPoint.cs`, and (newly) `.g.Main.cs`. Is that right?

Reply:
Yes, I think so.

Reply:
done

Reply:
Hm, on the name collision, I believe there indeed isn't a way to stick with both the files that are generated based on Q# files, and any files we generate additionally have the same .g.cs ending without the possibility that they collide. What we could do is have a convention but two endings as needed like .g1.cs and .g2.cs (etc) and edit the sdk to look for that pattern instead.
Rather than applying that strategy in a more ad-hoc manner and giving the entry point related files different names, I suggest to make it more general and track during emission which file names are already used, and increment a counter as needed. It would be a good opportunity to fix this bug: https://github.com/microsoft/qsharp-runtime/issues/357

Reply:
Discussed offline. Here is the strategy agreed to so far:
 - `*.g.cs` for regular generated C# files.
 - `*.g.EntryPoint.cs` for generated entry point files.
 - `EntryPoint.g.Main.cs` for the generated Main file.
 - If when the system goes to write a file that already exists, it will append a '1' to the 'g' token of the file name, for example `*.g1.cs`. If that also already exists, it will increment that number as needed, `*.g2.cs`, `*.g3.cs` and so on. This goes for entry points as well: `*.g1.EntryPoint.cs` `*.g2.EntryPoint.cs` and so on. This will be applicable to Main also, but it should need it. It will probably just be easier to implement it without special-casing Main.

The reason to do both methods is that the incrementing strategy will avoid name conflicts in general, and the difference in suffixes will add additional structure and clarity to the generated files.

Reply:
@bettinaheim The work to implement the enumeration in the file names (bullet 4 above) is proving to be too complicated a piece of work for it to be feasible to merge in with the rest of the work for Multiple Entry Points. Therefore I am dropping the last bullet off from the list and #357 will have to be addressed separately. That work really is unrelated to Multiple Entry Points, so it would be better to have a separate PR for that anyway.